WalletWasabi/WabiSabi/Models/Decomposition/CombinationsOfASize.cs

```diff
@@ -0,0 +1,180 @@
+
+		// Prune an array of decompositions, ensuring that the largest output in
+		// the remaining range is greater than LargestOutput.
+		private static ReadOnlyMemory<Decomposition> PruneLexicographically(ReadOnlyMemory<Decomposition> orderedDecompositions, Money largetstOutputLowerBound)
+			=> orderedDecompositions[Range.EndAt(
+				FindIndex(
+					orderedDecompositions.Span,
+					new Decomposition(largetstOutputLowerBound - 1L),
+					new ReverseLexicographicalComparer()))];
+
+		// Prune an array of decompositions, restricting to a range of total
+		// values.
+		// This is a private method, made internal so that it has unit tests.
+		internal static ReadOnlyMemory<Decomposition> PruneByTotalValue(ReadOnlyMemory<Decomposition> byTotalValue, Money maximum, Money minimum)
+			=> byTotalValue[new Range(
+				FindIndex(byTotalValue.Span, maximum),
+				FindIndex(byTotalValue.Span, minimum - 1L))];
+
+		// Find an index for a given total value, or the insert where
+		// it would be inserted.
+		private static Index FindIndex(ReadOnlySpan<Decomposition> orderedDecompositions, Money target)
+		{
+			if (orderedDecompositions.Length == 0)
+			{
+				return 0;
+			}
+			else if (target >= orderedDecompositions[0].TotalValue)
+			{
+				return 0;
+			}
+			else if (target < orderedDecompositions[^1].TotalValue)
+			{
+				return ^0;
+			}
+			else
+			{
+				return FindIndex(orderedDecompositions, new Decomposition(target), new ReverseTotalValueComparer());
+			}
+		}
+
+		// Find an index by binary searching using a comparer.
+		private static Index FindIndex(ReadOnlySpan<Decomposition> orderedDecompositions, Decomposition prototype, IComparer<Decomposition> comparer)
+		{
+			var i = orderedDecompositions.BinarySearch(prototype, comparer);
+
+			if (i < 0)
+			{
+				return ~i;
+			}
+			else
+			{
+				// BinarySearch doesn't necessarily return the first entry, so
+				// do an additional backwards linear scan to find it.
+				while (i > 0 && comparer.Compare(orderedDecompositions[i - 1], orderedDecompositions[i]) == 0)
+				{
+					i--;
+				}
+
+				return i;
+			}
+		}
+
+		private class ReverseTotalValueComparer : Comparer<Decomposition>
+		{
+			// Compare only using the total value, ignoring the individual
+			// outputs, used in FindIndex. Note that the order is always
+			// descending, so the arguments are reversed.
+			public override int Compare(Decomposition? x, Decomposition? y)
+				=> (x, y) switch
+				{
+					(null, null) => 0,
+					(null, _) => 1,
+					(_, null) => -1,
+					({} right, {} left) =>	left.TotalValue.CompareTo(right.TotalValue)
+				};
+		}
+
+		private class ReverseLexicographicalComparer : Comparer<Decomposition>
+		{
+			// Compare only using the first (largest) output value.
+			public override int Compare(Decomposition? x, Decomposition? y)
+				=> (x, y) switch
+				{
+					(null, null) => 0,
+					(null, _) => 1,
+					(_, null) => -1,
+					({} right, {} left)=> y.Outputs[0].CompareTo(x.Outputs[0])
```

Comment:
if we don't use `left` and `right` then it would be better to remove them.

Reply:
I don't understand why, but this is a breaking change:

```suggestion
					_ => y.Outputs[0].CompareTo(x.Outputs[0])
```